pulser/waveforms.py

```diff
@@ -739,6 +739,108 @@ def __mul__(self, other: float) -> InterpolatedWaveform:
         )
 
 
+class KaiserWaveform(Waveform):
+    """A Kaiser window of a specified duration and beta parameter.
+
+    For more information on the Kaiser window and the beta parameter,
+    check the numpy documentation for the kaiser(M, beta) function:
+    https://numpy.org/doc/stable/reference/generated/numpy.kaiser.html
+
+    Args:
+        duration (int): The waveform duration (in ns).
+        area (float): The integral of the waveform. Can be negative,
+            in which case it takes the positive waveform and changes the sign
+            of all its values.
+        beta (Optional[float]): The beta parameter of the Kaiser window.
+            The default value is 14.
+    """
+
+    def __init__(
+        self,
+        duration: Union[int, Parametrized],
+        area: Union[float, Parametrized],
+        beta: Optional[Union[float, Parametrized]] = 14.0,
+    ):
```

Comment:
Cool! How do you feel about having `KaiserWaveform.from_max_val()` too?

Reply:
`BlackmanWaveform.from_max_val()` relies on this to compute a value of `duration` to start with:

```
        # A normalized Blackman waveform has an area of 0.42 * duration
        duration = np.ceil(area / (0.42 * max_val) * 1e3)  # in ns
```

I could not find anything similar for a Kaiser window, especially since `beta` has a huge impact.

The only way I can think of is to start with `duration = 1` and then increment it... Not very "elegant"...

Do you think of any other approach?

Reply:
Indeed, it's not feasible to integrate the Kaiser window. What I have in mind is not much better than what you suggest, but it can take advantage of a couple facts:
- For a fixed value of ß, the Kaiser window's integral is proportional to the duration of the waveform, which means there is some proportionality factor `k(ß)` such that the area `A` is `A = k * duration`. Conversely, `duration = A / k`. 
- We know this factor decreases monotonically with ß. It starts as `k(0) = 1`, then `k(5) = 0.54`, `k(6) = 0.5`, `k(8.3) = 0.42`, and so on... Essentially, the window slims down with increasing ß, and so the area will decrease for a fixed duration.

My proposal is to take the given `beta` and get the nearest know `k(ß)` factor obeying `ß < beta`. This will give an upper-bound on the area for a fixed duration and, conversely, a lower-bound for the `duration` for the given area, which we can use as a starting point. From there, it's a matter of gradually increasing the duration, like we do for the Blackman. We can either increase in steps of 1 to make sure we don't miss the sweet spot, or try a bigger step and then correct in case of an overshoot.

Also, the value of `k` decreases progressively slower as ß increases, so for large values of ß we won't be that far off if we start off with the value of `k(14) = 0.321`, which can be the `k` factor corresponding to the largest ß we care to store.

Reply:
Working (hard) on it. But quite busy lately 😓. I will write more and PR something ASAP.

Reply:
No rush at all, take your time. Thank you for your efforts! 😄

Reply:
Here we go again!

Out of curiosity, I've been working on some kind of dichotomic approach but yours if much wiser and more efficient.

I spent a lot of time exploring this method and here are a few conclusions.

The ratio `area / duration` fluctuates a lot on small durations (mostly because of the even/odd durations artifact already mentioned in #218) but it tends to stabilize  after a while.

Hence, we can compute a satisfying value, without having to store any of them, by first creating a disposable waveform of `duration = 100` with the given `beta`.

Using this ratio, the computed duration guess is perfect most of the time, with a few exceptions by +/- 2. For 100% success, we can start from `duration_guess - 2` and increment step by step, as we do for `BlackmanWaveform`.

But this is true only when the duration guess is above 10. Below, the seesawing effect is very strong and I recommend testing all waveforms for duration from 1 to 10 and selecting the best one.

I wrote the corresponding code. It works perfectly. But there is still room for some improvement.

What do you think?

Reply:
This sounds like a great solution, I'll take a look right now.

Reply:
When the duration guess is less than or equal to 10, all durations from 1 to 15 are tested (stopping at 10 proved to be a bad idea since a guess on 9 might yield an optimal value on 11).

Otherwise: I enhanced the path to the best duration. Starting from the guessed duration minus 2 was not always a good pick. I tried with 99 % of the guessed duration but it is not very efficient. So, I simply start from the guessed duration and decide if it needs to be incremented or decremented towards the optimal duration. This leads to as few loops as possible.

Note: Not tested yet for negative areas and I think it would fail...

Reply:
I didn't think of negative areas for the `BlackmanWaveform` either. The best solution is probably to use the absolute value and then set the appropriate sign on the final waveform.